lib/tpm2_eventlog_yaml.c

```diff
@@ -401,14 +401,44 @@ bool yaml_specid_callback(TCG_EVENT const *event, void *data) {
     return yaml_specid_event(event, count);
 }
 
+static void yaml_eventlog_pcrs(tpm2_eventlog_ctx_t *ctx) {
+
+    char hexstr[DIGEST_HEX_STRING_MAX] = { 0, };
+
+    tpm2_tool_output("pcrs:\n");
+
+    tpm2_tool_output("  sha1:\n");
+    for(unsigned i = 0 ; i < TPM2_MAX_PCRS ; i++) {
+        bytes_to_str(ctx->sha1_pcrs[i], sizeof(ctx->sha1_pcrs[i]),
+            hexstr, sizeof(hexstr));
+        tpm2_tool_output("    %2d : 0x%s\n", i, hexstr);
+    }
+
+    tpm2_tool_output("  sha256:\n");
+    for(unsigned i = 0 ; i < TPM2_MAX_PCRS ; i++) {
+        bytes_to_str(ctx->sha256_pcrs[i], sizeof(ctx->sha256_pcrs[i]),
+            hexstr, sizeof(hexstr));
+        tpm2_tool_output("    %2d : 0x%s\n", i, hexstr);
+    }
+}
+
 bool yaml_eventlog(UINT8 const *eventlog, size_t size) {
 
     size_t count = 0;
+    tpm2_eventlog_ctx_t ctx = {
+        .data = &count,
+        .specid_cb = yaml_specid_callback,
+        .event2hdr_cb = yaml_event2hdr_callback,
+        .digest2_cb = yaml_digest2_callback,
+        .event2_cb = yaml_event2data_callback,
```

Comment:
These seem static except for the test code, so why are we exposing them as fn pointers? Just call the function directly and you can use ld wrap for hijacking during unit tests.

I don't want to add indirection complexity unless its needed, and I see that Phillip already did, your just moving where they get set, but I don't see the value in either of them.

Reply:
I'm not sure how the test code could do the ld wrapping trick to hijack the function pointers since there are many different combinations used in the tests.  The busybox executable also introduces a problem with making these global functions since both the `tpm2_eventlog.c` and `tpm2_checkquote.c` programs want to call into the eventlog parser.

Using a streaming visitor pattern might be the wrong approach for this; perhaps it would be better to parse the event log and return a tree structure of some sort that the caller can then traverse on their own, rather than having dynamic callbacks. It's not like these logs are all that large.

Reply:
Ahh yeah, I grepped for parse_eventlog and I now see the differences in uses. This approach is fine.